flink-core/src/main/java/org/apache/flink/api/common/time/Deadline.java

```diff
@@ -47,6 +48,19 @@ public Duration timeLeft() {
 		return Duration.ofNanos(Math.subtractExact(timeNanos, System.nanoTime()));
 	}
 
+	/**
+	 * Returns the time left between the deadline and now. If no time is left, a {@link TimeoutException} will be thrown.
+	 *
+	 * @throws TimeoutException if no time is left
+	 */
+	public Duration timeLeftIfAny() throws TimeoutException {
+		long nanos = Math.subtractExact(timeNanos, System.nanoTime());
+		if (nanos <= 0) {
+			throw new TimeoutException();
+		}
+		return Duration.ofNanos(nanos);
```

Comment:
You haven't understood the problem.

Presume that hasTimeLeft() returns true because there's still 1 nanosecond left. Your subsequent call to timeLeft() may return a value < 0, because time doesn't stop. This results in you passing in a negative value to sleep() calls which is commonly rejected with things like IllegalArgumentException, when for clarity you want to fail with a TimeoutException.

Reply:
You could of course first call `timeLeft`, convert to nanos and check whether it's negative, but it seems quite obtuse.

Reply:
Oh yes my bad to misunderstand the problem.
Thanks for the explanation!